core/domain/exp_services.py

```diff
@@ -1991,21 +1998,38 @@ def get_exp_with_draft_applied(exp_id, user_id):
         user_id: str. The id of the user whose draft is to be applied.
 
     Returns:
-        Exploration. The exploration domain object.
+        Exploration or None. Returns the exploration domain object with draft
+        applied, or None if draft can not be applied.
     """
     exp_user_data = user_models.ExplorationUserDataModel.get(user_id, exp_id)
     exploration = get_exploration_by_id(exp_id)
     if exp_user_data:
         if exp_user_data.draft_change_list:
+            draft_change_list_exp_version = (
+                exp_user_data.draft_change_list_exp_version)
             draft_change_list = [
                 exp_domain.ExplorationChange(change)
                 for change in exp_user_data.draft_change_list]
+            if (
+                    exploration.version >
+                    exp_user_data.draft_change_list_exp_version):
+                logging.info(
+                    'Exploration and draft versions out of sync, trying '
+                    'to upgrade draft version to match exploration\'s.')
+                new_draft_change_list = (
+                    draft_upgrade_services.try_upgrading_draft_to_exp_version(
+                        draft_change_list,
+                        exp_user_data.draft_change_list_exp_version,
+                        exploration.version, exp_id))
+                if new_draft_change_list is not None:
+                    draft_change_list = new_draft_change_list
```

Comment:
Are we storing this new_draft_change_list in `ExplorationUserDataModel` anywhere?

Reply:
We're not storing the new_draft_change_list other than applying it on the Exploration.

Sean made a comment in https://github.com/oppia/oppia/pull/7036/#discussion_r298793073 about it's counterintuitive to change models on a get() request. 

new_draft_change_list will automatically get saved to ExplorationUserDataModel on the next autosave though.